pyfifotax/report_data.py

```diff
@@ -0,0 +1,535 @@
+        if not self.legacy_mode:
+            self.report_events.extend(
+                MoneyTransferEvent.from_report(raw_data.money_transfers)
+            )
+            self.report_events.extend(ESPPEvent.from_report(raw_data.espp))
+
+        self.report_years = set(
+            [
+                e.date.year
+                for e in self.report_events
+                if not isinstance(e, StockSplitEvent)
+            ]
+        )
+
+    def process_report_events(self):
+        for event in self.report_events:
+            if isinstance(event, RSUEvent):
+                self.held_shares[event.symbol].push(event.received_shares)
+
+                bonus = awv.AWVEntryZ4RSUBonus(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.received_shares.currency,
+                    value=event.received_shares.total_buy_value()
+                    + event.withheld_shares.total_buy_value(),
+                )
+                bought_shares = awv.AWVEntryZ10RSUDeposit(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.received_shares.currency,
+                    quantity=event.received_shares.quantity
+                    + event.withheld_shares.quantity,
+                    value=event.received_shares.total_buy_value()
+                    + event.withheld_shares.total_buy_value(),
+                )
+                withheld_shares = awv.AWVEntryZ10RSUTaxWithholding(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.received_shares.currency,
+                    quantity=event.withheld_shares.quantity,
+                    value=event.received_shares.total_buy_value(),
+                )
+                self.awv_z4_events.append(bonus)
+                self.awv_z10_events.append(bought_shares)
+                if event.withheld_shares.quantity > 0:
+                    self.awv_z10_events.append(withheld_shares)
+
+            elif isinstance(event, ESPPEvent):
+                self.held_shares[event.symbol].push(event.received_shares)
+
+                bonus = awv.AWVEntryZ4ESPPBonus(
+                    symbol=event.symbol,
+                    currency=event.currency,
+                    date=event.date,
+                    value=event.bonus,
+                )
+
+                bought_shares = awv.AWVEntryZ10ESPPDeposit(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.currency,
+                    quantity=event.received_shares.quantity,
+                    value=event.received_shares.total_buy_value(),
+                )
+
+                self.awv_z4_events.append(bonus)
+                self.awv_z10_events.append(bought_shares)
+
+            elif isinstance(event, DividendEvent):
+                self.held_forex[event.currency].push(event.received_net_dividend)
```

Comment:
...except for EUR (since it's not tracked).

This push into the EUR queue is not an error in itself but triggers logging and since pushing into EUR queue is not tracked elsewhere (but popping is) dividend events show a huge negative EUR number.

Reply:
i don't fully understand
technically, most EUR transactions are tracked nevertheless, however, you can buy shares from EUR or FOREX without having a positive balance initially
if there is an actual issue with negative EUR balances starting from a positive one, i can investigate

Reply:
No, there's no issue here. I just see something similar in the debug logs:

```
[INFO] Dividends were paid on 2020-11-25: 1.10 EUR
[DEBUG] [Σ -10.99 EUR] Pushing 1.10 EUR to the queue which was acquired on 2020-11-25
```

The dividend payment event was a place where a possible negative EUR balance was exposed through the logging. But now, yes, I realised that since at most other places they are handled, it's not wise to remove the handling from here.

Reply:
You can see the current logging implementation in #10 (which is based on your current commit). Add `--log_level DEBUG` to see all messages.

I'm already finished with the verification but there's something wrong with pop logging (`log_pop()`). This part doesn't always behave well:

> Popping -X EUR (as a part of Y EUR)

I'll continue debugging it on Monday and if I find that the problem is within logging, I'll consider this PR finished :).

Reply:
I think I found the cause: the EUR balance in `CurrencyConversionEventFromEURToForex` and `CurrencyConversionEventFromForexToEUR` events is not tracked. 

An example: add a foreign currency, convert it to EUR and withdraw EUR. Basically the usual Schwab use case. Adding EUR with the conversion is not tracked but the withdrawal is, and that causes the EUR balance to become negative eventually.

Reply:
oh right, sorry, I forgot that, there is literally a comment which says "not tracked" in the code ...
should we fix that or go with the negative balance for now?

Reply:
as it is probably quite an easy fix, i added a first version of tracking these things, too, based on the daily rate on that day